narwhal/worker/src/tests/handlers_tests.rs

```diff
@@ -4,72 +4,94 @@
 use test_utils::CommitteeFixture;
+use types::WorkerToWorkerServer;
 
 #[tokio::test]
 async fn synchronize() {
     telemetry_subscribers::init_for_testing();
 
     let (tx_synchronizer, _rx_synchronizer) = test_utils::test_channel!(1);
     let (tx_primary, _rx_primary) = test_utils::test_channel!(1);
-    let (tx_request_batches_rpc, mut rx_request_batches_rpc) = test_utils::test_channel!(1);
     let (tx_batch_processor, mut rx_batch_processor) = test_utils::test_channel!(1);
 
     let fixture = CommitteeFixture::builder().randomize_ports(true).build();
     let worker_cache = fixture.shared_worker_cache();
+    let name = fixture.authorities().next().unwrap().public_key();
     let id = 0;
 
     // Create a new test store.
     let store = test_utils::open_batch_store();
 
     let handler = PrimaryReceiverHandler {
+        name,
         id,
         worker_cache,
         store,
+        request_batches_timeout: Duration::from_secs(999),
         request_batches_retry_nodes: 3, // Not used in this test.
         tx_synchronizer,
-        tx_request_batches_rpc,
         tx_primary,
         tx_batch_processor,
     };
 
     // Set up mock behavior for child RequestBatches RPC.
     let target_primary = fixture.authorities().nth(1).unwrap();
-    let target_worker = target_primary.worker(id);
-    let target_worker_network_key = target_worker.info().name.clone();
     let target = target_primary.public_key();
     let batch = test_utils::batch();
     let missing = vec![batch.digest()];
     let message = WorkerSynchronizeMessage {
         digests: missing.clone(),
         target,
     };
-    let responder_handle = tokio::spawn(async move {
-        let (target, num_nodes, request, response_channel) =
-            rx_request_batches_rpc.recv().await.unwrap();
-        assert_eq!(target.unwrap(), target_worker_network_key);
-        assert!(num_nodes.is_none());
-        assert_eq!(request, WorkerBatchRequest { digests: missing });
-
-        // Reply with batch.
-        assert!(response_channel
-            .send(Ok(anemo::Response::new(WorkerBatchResponse {
-                batches: vec![batch.clone()]
-            })))
-            .is_ok());
-
-        // Ensure the handler sends the returned batch to the processor.
-        let recv_batch = rx_batch_processor.recv().await.unwrap();
-        let recv_digest = &recv_batch.digest();
-        debug!("Received batch {recv_digest:?} from handler for processing");
-        assert_eq!(recv_batch, batch);
-    });
+
+    struct MockWorkerToWorker {
+        expected_request: WorkerBatchRequest,
+        response: WorkerBatchResponse,
+    }
+    #[async_trait]
+    impl WorkerToWorker for MockWorkerToWorker {
+        async fn send_message(
+            &self,
+            _request: anemo::Request<WorkerMessage>,
+        ) -> Result<anemo::Response<()>, anemo::rpc::Status> {
+            unimplemented!();
+        }
+        async fn request_batches(
+            &self,
```

Comment:
When you said this was unwieldy i was expecting this to be much more involved than it appears to actually be. This actually looks fairly straight forward.

Reply:
It could certainly be worse! IMO it's still significantly longer than the code it replaces, plus now you have to look in three places (the mock impl, the values used to construct the mock impl, and the recv channel) to piece together what the mock behavior is, instead of the one linear responder task we had before. of course this is testing more—it's using the actual network instead of just stubbing it out at the channel. but i hope could still get closer towards how gmock for grpc looks.